scripts/test_e2e.sh

```
@@ -0,0 +1,144 @@
+#   E2E_THREADS   test threads for the merged binaries (default: ~1.5x cores)
+#   E2E_RETRIES   serial retry rounds for failed tests   (default: 1, 0=off)
+#   E2E_TIMEOUT   whole-run timeout in seconds           (default: 600, 0=off)
+#
+# NOTE: the build step (`cargo test --no-run`) writes to the default `target/`,
+# which rust-analyzer may lock while it (re)compiles after an edit — the
+# "==> Building" line can sit for a while in that case. It is the editor's
+# lock, not a hang.
+
+set -uo pipefail
+
+cd "$(git rev-parse --show-toplevel)"
```

Shellcheck (0.11.0)

[warning] Use 'cd ... || exit' or 'cd ... || return' in case cd fails.

(SC2164)